pkg/verify/signed_entity.go

```diff
@@ -237,27 +241,90 @@ func WithArtifactDigest(algorithm string, artifactDigest []byte) PolicyOptionCon
 	}
 }
 
-// Verify checks the cryptographic integrity of a given SignedEntity according
+// VerifyWithArtifact checks the cryptographic integrity of a given SignedEntity,
+// and enforces that the SignedEntity being verified:
+// 1. was created from, or references, the provided artifact, and
+// 2. was created by one of the provided CertificateIdentities.
+//
+// Verification will fail if the SignedEntity:
+// - does not have a certificate,
+// - does not have a valid signature, or was signed outside of the cert's
+//	 validity period
+//
+// If and only if the verification is successful, this function returns a
+// VerificationResult whose contents have been verified and which can be used
+// for additional policy enforcement (i.e. enforcing the presence of a given
+// provenance predicate).
+//
+// For a more convenient way to initialize a CertificateIdentity, consult the
+// NewShortCertificateIdentity function.
+//
+// For more information, consult: VerifyUnsafe, WithArtifact, WithCertificateIdentity
+func (v *SignedEntityVerifier) VerifyWithArtifact(entity SignedEntity, artifact io.Reader, identities ...CertificateIdentity) (*VerificationResult, error) {
+	policies := []PolicyOptionConfigurator{WithArtifact(artifact)}
+
+	for _, certID := range identities {
```

Comment:
>I think it's confusing to have so many ways of specifying CertificateIdentity. Should the user call verify.WithCertificateIdentity, or pass it in to Verify* (or both?)

🤔 Users should call `VerifyWithArtifact` or `VerifyWithArtifactDigest`. `VerifyWith*` do not accept `PolicyOptionConfigurator`s.

There is only one _safe_ way to specify a certificate identity, and it is by passing a slice into one of those two functions. 

>instead of having Verify* functions, go back to having one Verify function, but have it take arguments like insecureWithoutArtifactIsOkay or insecureWithoutCertificateIdentityIsOkay

Two points here:

1. I kinda like a distinction between "Insecure" which to my mind means "this can never be made safe", and "unsafe" which means "use with caution". You should NEVER use `WithoutAnyObserverTimestampsInsecure()` unless you're writing a test. But `VerifyUnsafe` just requires expertise and paranoia.
2. I like the simplicity of required arguments. I like how we've eliminated choices: you use one func, or the other. I feel like it may also simplify argument validation. You check if the `certificateIdentities` slice is empty in the Safe functions, and can avoid double negatives in the Unsafe functions.

---

In this PR I proposed:
1. `VerifyUnsafe`, `VerifyWithArtifact`, `VerifyWithArtifactDigest`?

but there is also the option of:

2. `Verify(entity, [{WithoutArtifactUnsafe || WithArtifactDigest || WithArtifact}, {WithoutCertificateIdentityUnsafe() || WithCertificateIdentity…}]`

Which one requires more effort from the developer to use properly? Which option is more likely to hide a footgun?

I'm still inclined to think that option 1 is slightly more straightforward, but on reflection option 2 is also fine & meets the criteria.

Reply:
Would it make sense to consider an option 3, where the artifact and identity options are required positional arguments?

`Verify(SignedEntity, ArtifactMatcher, IdentityMatcher)`, where `ArtifactMatcher` and `IdentityMatcher` are interface types that the verifier will call into to "match" the artifact in question. And we can provide a few types that serve as unsafe no-op alternatives, such as `UnsafeAnyArtifact` and `UnsafeAnyCertificateIdentity`.

Reply:
Yes, and, what if they're just a positional `ArtifactPolicyConfigurator` / `IdentityPolicyConfigurator`, so we don't have to change any of the existing logic / add new interfaces?

i.e.

```
type ArtifactPolicyConfigurator func(*PolicyOptions) error
```

and 

```
type IdentityPolicyConfigurator func(*PolicyOptions) error
```

and the existing logic stays the same.

Reply:
hrm maybe it'd be like,

`Verify(entity, ArtifactPolicyConfigurator, IdentityPolicyConfigurator…)`

so you can be like,

`Verify(entity, WithArtifact(), WithCertIdentity(), WithCertIdentity()…)`

The only thing it forecloses is the ability to add new policy enforcement options in the future.

Reply:
If we care for extensibility, for not have an options struct? Very simple to work with, and can easily extend the struct in the future without braking the API (i.e parameters added/removed). That is if course as long as the struct can be created with sensible defaults. I'm personally not a fan of the functional config pattern, even though I know it's popular in the Go world.

But @phillmv 's latest suggestion
```
Verify(entity, WithArtifact(), WithCertIdentity(), WithCertIdentity()…)
```

Looks good I think.

Reply:
Had a little chat with @kommendorkapten, and maybe an options struct is the way. Something like:

```go
id := NewShortCertificateIdentity("foo", "", "", "")
policy := NewPolicy(WithArtifact(foo), WithId(id), WithId(id2))
Verify(entity, trustMaterial, policy)
```

This would preserve the option of extending policy in the future, for example, maybe adding the ability to enforce a predicate type?

Barring any objections, I think I'll try whipping this up!

Reply:
OK, I've gone ahead and implemented this.